custom_image_utils/shell_script_generator.py

```diff
@@ -307,16 +308,36 @@ def _init_args(self, args):
         "--metadata=shutdown-timer-in-sec={shutdown_timer_in_sec},"
         "custom-sources-path={custom_sources_path}"
     )
+    if self.args["zone"]:
+      region = "-".join(self.args["zone"].split("-")[:-1])
+      metadata_flag_template += ',dataproc-region="{}"'.format(region)
     if self.args["optional_components"]:
-      optional_components = self.args["optional_components"].replace(',', '.')
-      metadata_flag_template += ',optional-components="{}"'.format(optional_components)
+      optional_components = self.args["optional_components"].split(',')
+      # convert to component names used inside image and join to set as metadata value
+      optional_image_components = '.'.join(self._get_optional_to_image_components(optional_components))
+      metadata_flag_template += ',optional-components="{}"'.format(optional_image_components)
     if self.args["dataproc_version"]:
       dataproc_version = self.args["dataproc_version"]
-      metadata_flag_template += ',dataproc-version="{}"'.format(dataproc_version)
+      metadata_flag_template += ',dataproc_dataproc_version="{}"'.format(dataproc_version)
     if self.args["metadata"]:
       metadata_flag_template += ",{metadata}"
     self.args["metadata_flag"] = metadata_flag_template.format(**self.args)
 
+  def _get_optional_to_image_components(self, optional_components):
+    """Get the equivalent component names in the image for user provided optional components."""
+    # Add new component here, if component name inside image scripts is different.
+    optional_to_image_component_map = {
+      "DOCKER": "DOCKER-CE",
+      "HIVE_WEBHCAT": "HIVE-WEBHCAT-SERVER",
+      "SOLR": "SOLR-SERVER",
+    }
+    optional_image_components = []
+    for component in optional_components:
+      image_component = optional_to_image_component_map.get(component, component)
```

Comment:
second `component` here is the value to use if the key is not found in the dictionary

Reply:
Yes. this is expected.
This conversion is only required for some components, which have different names visible to user compared to internal name used in the scripts.
For ex: `DOCKER-CE` is the internal name, where as user just passes `DOCKER` for optional component.

So, if the entry is not present in this special map, then use the component name whatever passed from the user. There is a separate validation of user input for valid list of optional components. (in args_parser).